### PR TITLE
Update SlackApi

### PR DESCRIPTION
**Pull request by janboll**
* Minor refactoring: move calls to _get into seperate methods for easier
  readability
* Refactor slack api creation, remove init_slack functions, create class
  functions for instance creation

Refactoring improvements where discovered while working on: https://issues.redhat.com/browse/APPSRE-3988

### CODE REVIEW

File: reconcile/utils/slack_api.py [hunk 1]
```diff
@@ -285,22 +292,28 @@ def get_user_id_by_name(self, user_name: str) -> str:
 
         return result['user']['id']
 
-    def get_channels_by_names(self, channels_names):
-        return {k: v['name'] for k, v in self._get('channels').items()
+    def get_channels_by_names(self, channels_names: str):
```
Comment (Piojo) [inline reconcile/utils/slack_api.py:295]: Type hint the return value, please: `-> dict:`

File: reconcile/utils/slack_api.py [hunk 2]
```diff
@@ -285,22 +292,28 @@ def get_user_id_by_name(self, user_name: str) -> str:
...
+        return {k: v['name'] for k, v in self.get_channels()
                 if v['name'] in channels_names}
 
-    def get_channels_by_ids(self, channels_ids):
-        return {k: v['name'] for k, v in self._get('channels').items()
+    def get_channels_by_ids(self, channels_ids: str):
```
Comment (Piojo) [inline reconcile/utils/slack_api.py:299]: Is this a string? I'd expect some `Iterable[str]` given the name of the channel and the arguments. Also, type hint for the return value.

File: reconcile/utils/slack_api.py [hunk 3]
```diff
@@ -342,3 +355,48 @@ def _get(self, resource: str) -> Dict[str, Any]:
 
         self._results[resource] = results
         return results
+
+    @classmethod
+    def create_using_queries(cls, integration_name: str,
```
Comment (Piojo) [inline reconcile/utils/slack_api.py:360]: Wouldn't `create_from_queries` be more consistent with `create_from_dict`?

File: reconcile/utils/slack_api.py [hunk 4]
```diff
@@ -285,22 +292,28 @@ def get_user_id_by_name(self, user_name: str) -> str:
...
-    def get_users_by_ids(self, users_ids):
-        return {k: v['name'] for k, v in self._get('users').items()
+    def get_users_by_ids(self, users_ids: str):
+        return {k: v['name'] for k, v in self.get_users()
                 if k in users_ids}
 
+    def get_channels(self):
+        return self._get(self.RESOURCE_GET_CHANNELS).items()
+
+    def get_users(self):
```
Comment (Piojo) [inline reconcile/utils/slack_api.py:314]: These return `dict_items`.

**Comment by maorfr** [comment]
first round from me:
the new methods introduced to SlackApi to initiate from dict or from a query are creating a hard dependency between our own graphql schema and the slack logic.
in my view, the api/clients under `utils` should be as decoupled as possible from our schema.
i'd recommend to move this logic to slack_base perhaps?

i'm sure we are not following this 100% right now, but still worth the thought.

wdyt?

**Comment by janboll** [comment]
Thanks for the feedback, I'll move them back to  `slack_base.py`. I didn't saw the need for this file it, but I get it now with your  point of not querying server from the `utils` package.

**Comment by janboll** [comment]
> While I understand the intent of saving code and repetition, I find this new API confusing: some methods return dictionaries, some others return items to be iterated with.

Thanks for the feedback. There's no use case for the two new methods outside the class, since that I hid them.

File: reconcile/utils/slack_api.py (hunk 3 above)
Comment (janboll) [reply reconcile/utils/slack_api.py]: --> moved to slack_base.py and renamed

File: reconcile/openshift_upgrade_watcher.py [hunk 5]
```diff
@@ -31,7 +31,7 @@ def run(dry_run, thread_pool_size=10, internal=None, use_jump_host=True,
     )
 
     if not dry_run:
-        slack = init_slack_workspace(QONTRACT_INTEGRATION)
+        slack = SlackApi.create_using_queries(QONTRACT_INTEGRATION)
```
Comment (maorfr) [inline reconcile/openshift_upgrade_watcher.py:34]: still needs changing?

File: tools/qontract_cli.py [hunk 6]
```diff
@@ -857,7 +857,7 @@ def slack_usergroup(ctx, workspace, usergroup, username):
     Use an org_username as the username.
     To empty a slack usergroup, pass '' (empty string) as the username.
     """
-    slack = init_slack_workspace('qontract-cli')
+    slack = SlackApi.create_using_queries('qontract-cli')
```
Comment (maorfr) [inline tools/qontract_cli.py:860]: still needs changing?

File: reconcile/utils/slack_api.py [hunk 7]
```diff
@@ -252,7 +259,7 @@ def update_usergroup_users(self, id: str,
                 raise
 
     def get_random_deleted_user(self):
-        for user_id, user_data in self._get('users').items():
+        for user_id, user_data in self.get_users():
```
Comment (maorfr) [inline reconcile/utils/slack_api.py:262]: i don't see the `get_users` method... am i missing it?

Comment (janboll) [reply reconcile/utils/slack_api.py:262]: Let me add a test for this...

**Pull request closed without merging**